src/yadism/coefficient_functions/__init__.py

```diff
@@ -44,58 +44,44 @@ def __init__(self, esf):
         self.obs_name = esf.info.obs_name
         self.nf = nf_default(esf.Q2, esf.info.threshold)
         self.target = esf.info.target
+        self.scheme = esf.info.scheme
 
     def collect(self):
         comps = []
-
         family = self.obs_name.flavor_family
-
         # Adding light component
         if family in ["light", "total"]:
             comps.append(self.light_component())
-        if family == "heavy":
-            #  the only case in which an heavy contribution is not present in those
-            #  accounted for in total, it's whene heavy already became heavylight
+        if family == "heavy" and self.scheme != "FFN0":
+            # the only case in which an heavy contribution is not present in those
+            # accounted for in total, it's when heavy already became heavylight
             comps.extend(self.heavylight_components())
         if family in ["heavy", "total"]:
             comps.extend(self.heavy_components())
-
         return comps
 
     def light_component(self):
         nf = self.nf
         masses = self.masses
 
         comp = Component(0)
-
-        # the first condition essentially checks nf != 3
-        if nf in masses and masses[nf]:
-            nl = nf - 1
-            comp.extend(
-                fonll.kernels.generate_light(
-                    self.esf,
-                    nl,
-                    self.esf.info.theory["pto_evol"],
-                    self.esf.info.theory["pto"],
-                )
-            )
-            comp.extend(heavy.kernels.generate_missing(self.esf, nl, nl + 1))
-            comp.extend(
-                self.damp_elems(
-                    nl,
-                    fonll.kernels.generate_light_diff(
-                        self.esf,
-                        nl,
-                        self.esf.info.theory["pto_evol"],
-                    ),
-                )
-            )
-        else:
-            comp.extend(light.kernels.generate(self.esf, nf))
-
+        # light does not contain any mass effects and so is just always there
+        comp.extend(light.kernels.generate(self.esf, nf))
+        # instead missing encodes mass effects and so we need to fork:
         for ihq in range(nf + 1, 7):
             if masses[ihq]:
-                comp.extend(heavy.kernels.generate_missing(self.esf, nf, ihq))
+                if self.scheme == "FFN0":
+                    comp.extend(
+                        asy.kernels.generate_missing_asy(
+                            self.esf,
+                            nf,
+                            ihq,
+                            self.esf.info.theory["pto_evol"],
+                        )
+                    )
+                else:
+                    comp.extend(heavy.kernels.generate_missing(self.esf, nf, ihq))
+
         return comp
 
     def heavylight_components(self):
```

Comment:
to open another non-trivial point: since the whole point of NumFONLL is to account for simultaneous mass-effects we could/should make also heavylight more complicated and account for "higher"-mass effects: i.e. bottom-mass effects to F2c in FFNS4. This can be done using the `icoupl` argument of [`generate_missing`](https://github.com/NNPDF/yadism/blob/bc4ec4c9730724be453fea3e63d751b32cf3223a/src/yadism/coefficient_functions/heavy/kernels.py#L79) (and similar for asy)

Reply:
The goal of numerical FONLL is to simplify, not to complicate, definitely.

However, if we're not supporting it yet (it's not one of the features coming out of the box), open a dedicated issue. I would not include more features in this PR.

Reply:
Thanks!